File: wikipedia-fetcher.js
const axios = require('axios');

class WikipediaFetcher {
    constructor() {
        this.baseUrl = 'https://en.wikipedia.org/w/api.php';
    }

    async fetchPageLinks(pageTitle) {
        try {
            let allLinks = [];
            let continueToken = null;
            let iterations = 0;
            
            do {
                const params = {
                    action: 'query',
                    format: 'json',
                    titles: pageTitle,
                    prop: 'links',
                    pllimit: 'max', // Get maximum links per request (500)
                    origin: '*'
                };
                
                if (continueToken) {
                    params.plcontinue = continueToken;
                }

                const response = await axios.get(this.baseUrl, { params });
                const pages = response.data.query.pages;
                const pageId = Object.keys(pages)[0];
                
                if (pageId === '-1') {
                    console.error(`Page not found: ${pageTitle}`);
                    return [];
                }

                const links = pages[pageId].links || [];
                allLinks = allLinks.concat(links.map(link => link.title));
                
                // Check for continuation
                continueToken = response.data.continue ? response.data.continue.plcontinue : null;
                iterations++;
                
                // Small delay to avoid rate limiting
                if (continueToken) {
                    await new Promise(resolve => setTimeout(resolve, 100));
                }
                
                console.log(`  Fetched batch ${iterations} for ${pageTitle}: ${links.length} links`);
            } while (continueToken);
            
            console.log(`  Total links for ${pageTitle}: ${allLinks.length}`);
            return allLinks;
        } catch (error) {
            console.error(`Error fetching links for ${pageTitle}:`, error.message);
            return [];
        }
    }

    async fetchMultiplePages(pageTitles) {
        const results = {};
        
        for (const title of pageTitles) {
            console.log(`Fetching links from: ${title}`);
            const links = await this.fetchPageLinks(title);
            results[title] = links;
            
            // Add a small delay to avoid rate limiting
            await new Promise(resolve => setTimeout(resolve, 500));
        }
        
        return results;
    }
}

module.exports = WikipediaFetcher;

File: wikipedia_fetcher.py
import requests
import time

class WikipediaFetcher:
    def __init__(self):
        self.base_url = 'https://en.wikipedia.org/w/api.php'
    
    def fetch_page_links(self, page_title):
        """Fetch all links from a Wikipedia page."""
        all_links = []
        continue_token = None
        iterations = 0
        
        while True:
            params = {
                'action': 'query',
                'format': 'json',
                'titles': page_title,
                'prop': 'links',
                'pllimit': 'max',  # Get maximum links per request (500)
                'origin': '*'
            }
            
            if continue_token:
                params['plcontinue'] = continue_token
            
            try:
                response = requests.get(self.base_url, params=params)
                data = response.json()
                
                pages = data['query']['pages']
                page_id = list(pages.keys())[0]
                
                if page_id == '-1':
                    print(f"Page not found: {page_title}")
                    return []
                
                links = pages[page_id].get('links', [])
                all_links.extend([link['title'] for link in links])
                
                iterations += 1
                print(f"  Fetched batch {iterations} for {page_title}: {len(links)} links")
                
                # Check for continuation
                if 'continue' in data:
                    continue_token = data['continue']['plcontinue']
                    time.sleep(0.1)  # Small delay to avoid rate limiting
                else:
                    break
                    
            except Exception as e:
                print(f"Error fetching links for {page_title}: {str(e)}")
                return all_links
        
        print(f"  Total links for {page_title}: {len(all_links)}")
        return all_links
    
    def fetch_multiple_pages(self, page_titles):
        """Fetch links for multiple Wikipedia pages."""
        results = {}
        
        for title in page_titles:
            print(f"Fetching links from: {title}")
            links = self.fetch_page_links(title)
            results[title] = links
            time.sleep(0.5)  # Delay between different pages
        
        return results